File: baekjoon/trie/python/BOJ5052.py
from sys import stdin
from collections import defaultdict
readline = stdin.readline

class TrieNode:
    def __init__(self):
        self.word = False
        self.children = defaultdict(TrieNode)

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root

        for char in word:
            if node.word:
                return False
            
            node = node.children[char]
        
        if node.word:
            return False
        
        node.word = True
        return True


T = int(readline())

def solve():

    for _ in range(T):
        N = int(readline())
        numbers = list(sorted([readline().strip() for _ in range(N)], key=lambda x: len(x)))
        check_consistency(numbers)

def check_consistency(numbers):
    trie = Trie()

    for number in numbers:
        if not trie.insert(number):
            print("NO")
            return
        
    print("YES")

solve()

File: 0_1.baekjoon/trie/java/BOJ5052.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class BOJ5052 {
    static class TrieNode {
        boolean word;
        Map<Character, TrieNode> children;
        TrieNode() {
            word = false;
            children = new HashMap<>();
        }
    }

    static class Trie {
        TrieNode root;
        Trie() {
            root = new TrieNode();
        }

        boolean insert(String word) {
            TrieNode node = root;
            
            for (char ch : word.toCharArray()) {
                if (node.word)
                    return false;

                node.children.putIfAbsent(ch, new TrieNode());
                node = node.children.get(ch);
            }

            if (node.word)
                return false;

            node.word = true;
            
            return true;
        }
    }
    public static void main(String[] args) throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());

        for (int i=0;i<T;i++) {
            int N = Integer.parseInt(br.readLine());
            
            String[] numbers = new String[N];
            for (int j=0;j<N;j++) {
                numbers[j] = br.readLine();
            }

            Arrays.sort(numbers, (s1, s2) -> s1.length() - s2.length());
            
            checkConsistency(numbers);
            
        }
    }
    private static void checkConsistency(String[] numbers) {
        Trie trie = new Trie();

        for (String number : numbers) {
            if(!trie.insert(number)) {
                System.out.println("NO");
                return;
            }
        }
        System.out.println("YES");
    }
}